### PR TITLE
NEW: @W-18161190@: Make v5 the default, add in config-file option, and telemetry

### PR DESCRIPTION
**Pull request by stephen-carter-at-sf**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -150,22 +145,32 @@
...
+                        "description": "(Pilot) Discover critical problems and performance issues in your Apex code with ApexGuru, which analyzes your Apex files for you. This feature is in a closed pilot; contact your account representative to learn more."
+                    },
+                    "codeAnalyzer.useV4Deprecated": {
+                        "type": "boolean",
+                        "default": false,
+                        "markdownDescription": "Use Code Analyzer v4 (Deprecated) instead of Code Analyzer v5.\n\nWe no longer provide support for Code Analyzer v4.x and will soon be removing this setting, so we highly recommend you use [Code Analyzer v5.x](https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/code-analyzer.html) instead. Selecting this setting will ignore the Code Analyzer v5 settings and use the v4 settings instead.\n\nCreate an [issue](https://github.com/forcedotcom/sfdx-code-analyzer-vscode/issues) if you are having trouble switching to v5.x."
```
Comment (randi274) [inline package.json:153]: Nice! Let's hope someone comes to us looking for some migration help 🤞

File: src/extension.ts [hunk 2]
```diff
@@ -93,7 +93,15 @@ export async function activate(context: vscode.ExtensionContext): Promise<SFCAEx
     // =================================================================================================================
     // ==  Code Analyzer Run Functionality
     // =================================================================================================================
-    await establishVariableInContext('sfca.codeAnalyzerV4Enabled', () => Promise.resolve(!settingsManager.getCodeAnalyzerV5Enabled()));
+    await establishVariableInContext('sfca.codeAnalyzerV4Enabled', () => Promise.resolve(settingsManager.getCodeAnalyzerUseV4Deprecated()));
+
+    // Monitor the "codeAnalyzer.useV4Deprecated" setting with telemetry
```
Comment (randi274) [inline src/extension.ts:98]: ❤️

File: src/lib/scanner-strategies/scanner-strategy.ts [hunk 3]
```diff
@@ -13,11 +13,17 @@ export interface ScannerStrategy {
 
 
 export abstract class CliScannerStrategy implements ScannerStrategy {
+    private isValidEnvironment?: boolean;
+
     public async validateEnvironment(): Promise<void> {
+        if (this.isValidEnvironment) {
```
Comment (jfeingold35) [inline src/lib/scanner-strategies/scanner-strategy.ts:19]: Question: Do we also want some way of caching a false response? As it currently stands, an invalid environment wouldn't exit early here.

File: src/lib/scanner-strategies/v5-scanner.ts [hunk 4]
```diff
@@ -46,72 +47,34 @@ export class CliScannerV5Strategy extends CliScannerStrategy {
...
-        if (potentiallyLongRunningRules.length > 0) {
-            await this.confirmPotentiallyLongRunningScan(potentiallyLongRunningRules);
-        }
-
         const resultsJson: ResultsJson = await this.invokeAnalyzer(filesToScan);
 
         return this.processResults(resultsJson);
     }
 
-    private async getLongRunningRules(): Promise<string[]> {
```
[the remark this answers is not in the document]
Comment (jfeingold35) [reply src/lib/scanner-strategies/v5-scanner.ts:69]: ...Fair enough. I'm not thrilled about this. It's an important piece of functionality for later when we start adding in SFGE stuff.

File: src/lib/scanner-strategies/scanner-strategy.ts (hunk 3 above)
Comment (jfeingold35) [reply src/lib/scanner-strategies/scanner-strategy.ts:19]: Is it possible that the user could re-break it after they've fixed it?

File: src/lib/scanner-strategies/v5-scanner.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply src/lib/scanner-strategies/v5-scanner.ts:69]: Sorry but I would have had to update this code with my changes here and I noticed there were no tests here. So without tests - I wouldn't confidently know if my changes worked. And I didn't want to waste time adding in tests for this since we don't have any actual use cases for it now.

It's simple enough to add back in when we get this point - at which time we can add in tests.

File: src/lib/scanner-strategies/scanner-strategy.ts (hunk 3 above)
Comment (stephen-carter-at-sf) [reply src/lib/scanner-strategies/scanner-strategy.ts:19]: Yes - but that is an extreme edge case... (where in worse case scenario the error occurs later on during ascan) and isn't worth the performance hit that we are taking here.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jshackell-sfdc) [reply package.json:153]: Also, I think we should change the label of this setting to:

**Code Analyzer: Use V4 (Deprecated)**

I _think_ the label doesn't currently include the parens. I don't see where this is set in this PR, so just adding a note here. Thx.

File: src/lib/messages.ts [hunk 5]
```diff
@@ -7,7 +7,7 @@
 export const messages = {
     noActiveEditor: "Unable to perform action: No active editor.",
     staleDiagnosticPrefix: "(STALE: The code has changed. Re-run the scan.)",
-    stoppingV4SupportSoon: "We plan to stop supporting v4.x of Code Analyzer in the coming months. We highly recommend that you start using v5.x, which is currently in Beta, by setting 'Code Analyzer: Enable V5' to true. For information on v5.x, see https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/code-analyzer.html.",
+    stoppingV4SupportSoon: "We no longer provide support for Code Analyzer v4.x and will soon remove it from this VS Code Extension. We highly recommend that you start using v5.x by unselecting the 'Code Analyzer: Use V4 Deprecated' setting. For information on v5.x, see https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/code-analyzer.html.",
```
[the remark this answers is not in the document]
Comment (jshackell-sfdc) [reply src/lib/messages.ts:10]: Also, if you DO change the label of the "Code Analyzer: Use V4 (Deprecated)" setting, update this message accordingly (just add the parens).

File: src/lib/messages.ts [hunk 6]
```diff
@@ -73,8 +73,8 @@ export const messages = {
     error: {
         analysisFailedGenerator: (reason: string) => `Analysis failed: ${reason}`,
         pmdConfigNotFoundGenerator: (file: string) => `PMD custom config file couldn't be located. [${file}]. Check Salesforce Code Analyzer > PMD > Custom Config settings`,
-        sfMissing: "To use this extension, first install Salesforce CLI `sf` or `sfdx` commands.",
-        sfdxScannerMissing: "To use this extension, first install `@salesforce/sfdx-scanner`.",
+        sfMissing: "To use this extension, first install the Salesforce CLI `sf` command.",
```
Comment (jshackell-sfdc) [inline src/lib/messages.ts:76]:
```suggestion
        sfMissing: "To use this extension, first install Salesforce CLI.",
```

File: src/lib/messages.ts [hunk 7]
```diff
@@ -73,8 +73,8 @@ export const messages = {
...
+        sfdxScannerMissing: "To use the 'Use v4 Deprecated' setting, you first must first install `@salesforce/sfdx-scanner`. We recommend instead that you use v5 and unselect the 'Use v4 Deprecated' setting.",
```
Comment (jshackell-sfdc) [inline src/lib/messages.ts:77]:
```suggestion
        sfdxScannerMissing: "To use the 'Code Analyzer: Use V4 Deprecated' setting, you first must first install the `@salesforce/sfdx-scanner` Salesforce CLI plugin. But we no longer support v4, so we recommend that you use v5 instead and unselect the 'Use V4 Deprecated' setting.",
```

Comment (jshackell-sfdc) [reply src/lib/messages.ts:77]: Ditto previous message: if you change the label, remember to add the parens (Code Analyzer: Use V4 (Deprecated))

**Review by jshackell-sfdc — CHANGES_REQUESTED** [review]

File: package.json [hunk 8]
```diff
@@ -150,22 +145,32 @@
...
+                    },
                     "codeAnalyzer.ruleSelectors": {
                         "type": "string",
                         "default": "Recommended",
-                        "markdownDescription": "Specifies the default set of rules to use when executing Code Analyzer v5 (Beta). Specify the rules using their name, engine name, severity level, tag, or a combination. Use commas for unions (such as \"Security,Performance\") and colons for intersections (such as \"pmd:Security\" or \"eslint:3\")."
+                        "markdownDescription": "The rule selector(s) for the set of rules to use when scanning your code with Code Analyzer v5.\n\nSelect rules using their name, engine name, severity level, tag, or a combination. Use commas for unions (such as \"Security,Performance\") and colons for intersections (such as \"pmd:Security\" or \"eslint:3\").\n\nThis setting is equivalent to the `--rule-selector` flag from the CLI. See [examples](https://developer.salesforce.com/docs/atlas.en-us.sfdx_cli_reference.meta/sfdx_cli_reference/cli_reference_code-analyzer_commands_unified.htm#cli_reference_code-analyzer_rules_unified)."
```
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply package.json:168]: How about using the same starting words that we use in the cli help and instead just say 
```
Selection of rules used to scan your code with Code Analyzer v5.
```
as the first sentence.

File: src/lib/messages.ts (hunk 7 above)
Comment (stephen-carter-at-sf) [reply src/lib/messages.ts]: Fixed. Thanks.

File: src/lib/messages.ts (hunk 6 above)
Comment (stephen-carter-at-sf) [reply src/lib/messages.ts]: Fixed. Thanks.

File: src/lib/messages.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply src/lib/messages.ts]: Fixed. Thanks.

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply package.json]: Fixed. Thanks.

File: package.json [hunk 9]
```diff
@@ -150,22 +145,32 @@
...
                 }
             },
             {
-                "title": "Code Analyzer v5 (Beta)",
+                "title": "Code Analyzer v5",
                 "properties": {
+                    "codeAnalyzer.configFile": {
+                        "type": "string",
+                        "default": "",
+                        "markdownDescription": "Path to a [Code Analyzer configuration](https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/config-custom.html) file used to customize the engines and rules.\n\nCode Analyzer has an internal default configuration for its rule and engine properties. If you want to override these defaults, you can either add a 'code-analyzer.yml' or 'code-analyzer.yaml' file at the root of your project or explicitly specify your configuration file path with this setting.\n\nThis setting is equivalent to the `--config` flag from the CLI. See [configuration schema](https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/config-toplevel.html)."
```
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply package.json]: Fixed. Thanks.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply package.json]: Fixed. Thanks.

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (jshackell-sfdc) [reply package.json]: Like it, thx

**Review by jshackell-sfdc — APPROVED** [review]

**Review by randi274 — APPROVED** [review]
A few suggestions, but approved!

File: package.json [hunk 10]
```diff
@@ -150,22 +145,32 @@
...
+                        "description": "(Pilot) Discover critical problems and performance issues in your Apex code with ApexGuru, which analyzes your Apex files for you. This feature is in a closed pilot; contact your account representative to learn more."
+                    },
+                    "codeAnalyzer.Use v4 (Deprecated)": {
+                        "type": "boolean",
+                        "default": false,
+                        "markdownDescription": "Use Code Analyzer v4 (Deprecated) instead of Code Analyzer v5.\n\nWe no longer support Code Analyzer v4 and will soon remove this setting. We highly recommend that you use [Code Analyzer v5](https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/code-analyzer.html) instead. Selecting this setting ignores the Code Analyzer v5 settings and uses the v4 settings instead.\n\nIf you have having trouble switching to v5, create an [issue](https://github.com/forcedotcom/sfdx-code-analyzer-vscode/issues)."
```
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply package.json:153]: Yeah good idea... won't do it here - maybe in a stretch goal PR later on.

File: package.json [hunk 11]
```diff
@@ -150,22 +145,32 @@
...
                         "default": "Recommended",
-                        "markdownDescription": "Specifies the default set of rules to use when executing Code Analyzer v5 (Beta). Specify the rules using their name, engine name, severity level, tag, or a combination. Use commas for unions (such as \"Security,Performance\") and colons for intersections (such as \"pmd:Security\" or \"eslint:3\")."
+                        "markdownDescription": "Selection of rules used to scan your code with Code Analyzer v5.\n\nSelect rules using their name, engine name, severity level, tag, or a combination. Use commas for unions (such as \"Security,Performance\") and colons for intersections (such as \"pmd:Security\" or \"eslint:3\").\n\nThis setting is equivalent to the `--rule-selector` flag of the CLI commands. See [examples](https://developer.salesforce.com/docs/atlas.en-us.sfdx_cli_reference.meta/sfdx_cli_reference/cli_reference_code-analyzer_commands_unified.htm#cli_reference_code-analyzer_rules_unified)."
                     }
                 }
             },
             {
-                "title": "Code Analyzer v4",
+                "title": "Code Analyzer v4 (Deprecated)",
```
[the remark this answers is not in the document]
Comment (stephen-carter-at-sf) [reply package.json:173]: If you do - VS code automatically sorts things and it ends up way below. So it's in the general section for that reason... and it is short lived so I'm not worried about it.
